types/src/validator_verifier.rs

```diff
@@ -159,54 +168,114 @@ impl ValidatorVerifier {
         }
     }
 
-    /// This function will successfully return when at least quorum_size signatures of known authors
-    /// are successfully verified. Also, an aggregated signature is considered invalid if any of the
-    /// attached signatures is invalid or it does not correspond to a known author. The latter is to
-    /// prevent malicious users from adding arbitrary content to the signature payload that would go
-    /// unnoticed.
-    pub fn verify_aggregated_struct_signature<T: CryptoHash + Serialize>(
+    // Generates a multi signature from partial signatures without actually verifying it.
+    pub fn generate_multi_signature(
         &self,
+        partial_signatures: &PartialSignatures,
+    ) -> Result<(MultiSignature, Option<PublicKey>), VerifyError> {
+        if partial_signatures.is_empty() {
+            // In case of empty signature list, we don't have anything to aggregate and verify
+            return Ok((MultiSignature::empty(), None));
+        }
+        let validator_bitmask = self
+            .address_to_validator_info
+            .iter()
+            .map(|(address, _)| partial_signatures.signatures().contains_key(address))
+            .collect::<Vec<_>>();
+        // Perform an optimistic aggregation of the signatures.
+        let aggregated_sig = bls12381::Signature::aggregate(
+            partial_signatures.signatures().values().cloned().collect(),
+        )
+        .map_err(|_| VerifyError::FailedToAggregateSignature)?;
+
+        // Verify the optimistically aggregated signature.
+        let mut pub_keys_to_agg = vec![];
+        for address in partial_signatures.signatures().keys() {
+            pub_keys_to_agg.push(
+                self.address_to_validator_info
+                    .get(address)
+                    .ok_or(VerifyError::UnknownAuthor)?
+                    .public_key(),
+            );
+        }
+
+        let aggregated_key = PublicKey::aggregate(pub_keys_to_agg)
+            .map_err(|_| VerifyError::FailedToAggregateSignature)?;
+        Ok((
+            MultiSignature::new(validator_bitmask, Some(aggregated_sig)),
+            Some(aggregated_key),
+        ))
+    }
+
+    pub fn generate_and_verify_multi_signature<T: CryptoHash + Serialize>(
```

Comment:
I'd also rename this to `generate_and_verify_multi_signature`

Reply:
You mean `aggregate_and_verify_multi_signature` ? :)